students/BrianNisonger/session7/html_render.py

```diff
@@ -1,18 +1,177 @@
-#!/usr/bin/env python3
-
 """
 A class-based system for rendering html.
 """
 
 
 # This is the framework for the base class
-class Element(object):
+class Element:
+    tag = "html"
+    indent = "    "
 
-    def __init__(self, content=None):
-        pass
+    def __init__(self, content=None, **kwargs):
+        if content == None:
+            self.content = []
+        else:
+            self.content = [content]
+        self.kwargs = kwargs
 
     def append(self, new_content):
-        pass
+        self.content.append(new_content)
+
+    def _open_tag(self, indent=indent):
+        open_tag = f'<{self.tag}'
+        for attributes in self.kwargs:
+            open_tag = open_tag + (
+                f' {attributes}="{self.kwargs[attributes]}"')
+        open_tag = open_tag + (">\n")
+        return open_tag
+
+    def _close_tag(self):
+        close_tag = f'</{self.tag}>'
+        return close_tag
+
+    def render(self, out_file, indent=indent):
+        self.indent = indent
+        tag_indent = self.indent
+        out_file.write(tag_indent)
+        out_file.write(self._open_tag())
+        for content in self.content:
+            if content != None:
+                try:
+                    content.render(out_file, indent="")
+                except AttributeError:
+                    out_file.write(content.strip())
+                out_file.write("\n")
+        out_file.write(tag_indent)
+        out_file.write(self._close_tag())
+
+
+class Html(Element):
+    tag = "html"
+
+    def render(self, out_file, indent=""):
+        self.indent = indent
+        out_file.write("<!DOCTYPE html>\n")
+        super().render(out_file, self.indent)
+
+
+class Body(Element):
+    tag = "body"
+    indent = ""
+
+
+class P(Element):
+    tag = "p"
+
+
+class Head(Element):
+    tag = "head"
+
+
+class OneLineTag(Element):
+    indent = ""
+
+    def append(self, new_content):
+        self.content.append(f' {new_content}')
+
+    def _open_tag(self):
```

Comment:
DRY!

(don't repeat yourself)

the whole point of defining this as a method on the base Element class is that you can use it everywhere else.